### DISCUSSION TITLE
Хакатон Demhack 7 // 29.09 - 01.10 // Задача Lantern

### DISCUSSION THREAD

**Discussion by deanna1234** [getlantern/lantern-forum-ru#8] [Ideas]
Приветствуем участников Demhack 7!  

Спасибо, что выбрали для работы в рамках хакатона предложенную нами задачу. Мы очень ценим ваш интерес и вклад членов коммьюнити в развитие технологий, способствующих борьбе с цензурой. С момента своего основания Lantern остается верен своей миссии - делать интернет свободным и доступным для всех. И в течение следующих трех дней мы будем рады работать над этим вместе с вами! 
 
**Задача Lantern следующая:** 
Lantern Desktop поддерживается дистрибутивами Linux на базе Debian. Он может работать на других дистрибутивах (Fedora, Arch, Alpine, Void и т. д.), но его необходимо правильно упаковать для этих систем (RPM для Fedora, AUR для Arch и т. д.). Требуется: проанализировать пакеты, выпущенные Lantern для Debian, и создать инструменты сборки для упаковки приложения для различных версий Linux.

С нашей стороны с командой будет работать @reflog! Все уточнения, обсуждения и вопросы, просим команду оставлять в этой ветке форума. 

Всем продуктивной работы, скоро начинаем!

**Comment by glue0** [comment]
Добрый день!
Как ранее выяснили - будем работать только с бинарными файлами и deb пакетами . Сейчас активно занимаемся анализом библиотек и доставки в пакет
Можно узнать примерный список используемых библиотек с версиями?

**Reply by deanna1234** [reply]
@reflog ответь ребятам на вопрос выше, пожалуйста

**Reply by reflog** [reply]
@glue0 самый простой способ получить список - запустить `ldd lantern-desktop`

**Comment by glue0** [comment]
# Инструменты упаковки бинарных пакетов для различных версий Linux

# Description

Lantern Desktop поддерживается дистрибутивами Linux на базе Debian. Он может работать на других дистрибутивах (Fedora, Arch, Alpine, Void и т. д.), но его необходимо правильно упаковать для этих систем (RPM для Fedora, AUR для Arch и т. д.).

Задача: проанализировать пакеты, выпущенные Lantern для Debian, и создать инструменты сборки для упаковки приложения для различных версий Linux.

Тех задача: Требовалось перепаковать debian или бинарный файл (без исходников) для других дистрибутивов в автоматическом режиме. В качестве входных данных был дан пакет Lantern Desktop.

# Reference

## UNZIP

https://www.debian.org/doc/debian-policy/
`ar -x` - unpack on debian
Control (pack meta, hash sums) + Data
Unpack data -> usr: bin, doc
.tar.gz - `tar -xzf`
__Universal__: `dpkg-deb -e` :preinst, postinst, prerm, postrm

## PACKAGING

Options:

- docker

- manual packing and linking

- chroot

- __GitHub Actions__

__Selected way__: Github Actions

Package-specific applications:

1. Arch - AUR:

- `mpkg`
Conclusion - mpkg with linking
- `makepkg`

2. RPM - Fedora:

Classic `rpm-tools` - https://rpm-packaging-guide.github.io/
`alien`  - from deb to rpm : GH Actions/debian env
`rpkg` - build and linking [due](https://www.reddit.com/r/linuxquestions/comments/fag9ej/best_tutorial_for_automating_creation_of_deb_rpm/i)

3. Alpine:

https://wiki.alpinelinux.org/wiki/Creating_an_Alpine_package
`dh-make`

4. Debian:

`dpkg`

5. Nix

<https://nixos.wiki/wiki/Packaging/Binaries>

Conclusion:
`makepkg` - AUR
`rpkg` - RPM

[1] Deep Linking - https://wiki.archlinux.org/title/Creating_packages_for_other_distributions

## META ANALYSING

[PKG Format](http://refspecs.linux-foundation.org/LSB_3.2.0/LSB-Core-generic/LSB-Core-generic/pkgformat.html)

RPM
Fedora - RPM
<https://web.archive.org/web/20050312181634/http://www.hut.fi/~tkarvine/rpm-build-as-user.html>
<https://web.archive.org/web/20060219023154/http://www.rpm.org/max-rpm/rpm.8.html>


MISC:
[DPMD](https://github.com/FooBarWidget/debian-packaging-for-the-modern-developer)

# Process 

 SHEME

1. GitHub Actions

2. Scripts:

- Скачка, распаковка, верификация

- анализ, генерация

- *no needed* скачка, компляция библиотек

- сборка

3. Libraries bug

4. Сборка в Aur

MakePkg + Linking

## BUG

1. PKGBUILD not executed - missing libraries.
`libcap libappindicator-gtk3` - lib linking
Add Libraries to package, linked them manually.

## BACKLOG

RPM
Generating by:

- `rpkg` or `rpm-tools` on Fedora environment

- `alien` on Debian env


## QUESTIONS

1) Sources or binaries?
Binaries.

2) Dependences?
analysed - `libcap libappindicator-gtk3`

# SPECS

Package Meta
```
control
Package: lantern
Version: 7.4.0
License: Apache-2.0
Vendor: Brave New Software Project, Inc
Architecture: amd64
Maintainer: Lantern Team <team@getlantern.org>
Installed-Size: 38909
Section: net
Priority: extra
Homepage: https://lantern.io
Description: Censorship circumvention tool\nLantern allows you to access sites blocked by internet censorship.\nWhen you run it, Lantern reroutes traffic to selected domains through servers located where such domains are uncensored.
md5sums
2c8b054c8a12515f1a3837c0c29ccf31  usr/lib/lantern/.packaged-lantern.yaml
0584f926263bad2b8ff880d3c0fba7de  usr/lib/lantern/lantern-binary
0da640fa8b0776eac7d2a49fcc605058  usr/lib/lantern/lantern.sh
d784fa8b6d98d27699781bd9a7cf19f0  usr/lib/lantern/lantern.yaml
8e8eb5924c48cb868a4af4ea072cdec1  usr/share/applications/lantern.desktop
f5dc48790d23fef4d90fb5a036ed35bc  usr/share/doc/lantern/changelog.gz
0de8547331c4a8ea3b696716d9d8c050  usr/share/doc/lantern/copyright
f85508219e86affaa2cb80de1fec31e4  usr/share/icons/hicolor/128x128/apps/lantern.png
```

***
                                               
# DOC 

1. Init environment and choose target systems (with applications).

`.yaml` - build configuration for GH Actions [YAML](https://github.com/d3spairx/lantern-distro-port/blob/main/.github/workflows/main.yaml) ([doc](https://docs.github.com/ru/actions/examples/using-scripts-to-test-your-code-on-a-runner))

- `dpkg-deb` - extract deb packet

2. Build for different systems

- `makepkg` - build aur packet [AUR](https://github.com/d3spairx/lantern-distro-port/releases/tag/7.4.0)

[PKGBUILD](https://github.com/d3spairx/lantern-distro-port/blob/main/PKGBUILD) - build script with
+ metadata
- change desktop
- add lib link

Build RPM packet by (REF TO BACKLOG):
- `rpkg` [doc](https://linux.die.net/man/1/rpkg) or `rpm-tools`[doc](https://rpm-packaging-guide.github.io/#packaging-software) on Fedora environment
- `alien` on Debian env

**Comment by deanna1234** [comment]
@reflog Привет! Сориентируй, пожалуйста, когда сможешь оценить решение ребят и вернуться с комментариями/вопросами или уже сразу с ответом - справились они с задачей или пока нет?

**Comment by reflog** [comment]
@glue0 Привет! Может я не туда смотрю, но вот тут https://github.com/d3spairx/lantern-distro-port/blob/main/.github/workflows/main.yaml почти ничего нет.
Может вы в другом репозитории работали?

**Reply by glue0** [reply]
Модифицировали код - https://github.com/d3spairx/lantern-distro-port/tree/test
Для арча была сложность в правах для выполнения: в gh actions есть возможность запуска только в ubuntu. Поэтому запускали билд скрипт на локальном окружении для тестов.
Сейчас поправили [билд](https://github.com/d3spairx/lantern-distro-port/blob/test/.github/workflows/main.yaml) ( pkgbuild_archlinux-package-build ) для арча на gh a - он работет полностью.

Но для сборки других пакетов решили сделать упрощение, чтобы не запихивать в docker контейнеры.
Вообще обнаружилась проблема в запуске многих пакетов, поэтому в качестве лучшего решения сейчас - был выбран [nfpm](https://github.com/goreleaser/nfpm) c [доками](https://nfpm.goreleaser.com/)
Зарелизили порт c помощью nfpm в rpm, aur, apk
[Config](https://github.com/d3spairx/lantern-distro-port/blob/test/nfpm.yaml) и [новый билд](https://github.com/d3spairx/lantern-distro-port/blob/test/.github/workflows/main.yaml)
С rpm всё отлично. 
aur устанавливается немного ругаясь на права, но в целом работает. 
apk не билдится из-за немного неправильных прав на директориях и файлах. решается добавлением директив в конфиг nfpm

**Comment by reflog** [comment]
Проверка решения закончилась. С задачей вы справились! Молоды! Спасибо за участие
